### DISCUSSION TITLE
How to know if a driver DNF'd?

### DISCUSSION THREAD

**Discussion by KevinCamacho** [br-g/openf1#63] [Q&A]
I'm using the positions API to get the finishing positions for each driver for a race. If a driver DNFs, the API just returns a 1-20 final position. 

Bottas' positions from this past Chinese GP, for example. He was the first to DNF so came "20th"
https://api.openf1.org/v1/position?session_key=9673&driver_number=77

**Comment by br-g** [comment]
Hello,

Unfortunately, there's currently no simple method to identify a DNF. This feature might be added in the future, but it's not a current focus.

However, you can determine DNF by examining lap or interval data. For instance, Bottas' last lap in the Chinese Grand Prix was the 20th, even though the race had 57 laps:
[https://api.openf1.org/v1/laps?session_key=9673&driver_number=77](https://api.openf1.org/v1/laps?session_key=9673&driver_number=77)

Similarly, his last recorded gap to the race leader is listed as "37L" (37 laps behind):
[https://api.openf1.org/v1/intervals?session_key=9673&driver_number=77](https://api.openf1.org/v1/intervals?session_key=9673&driver_number=77)

**Reply by KevinCamacho** [reply]
I tried to do some investigation around what this kind of solutino would look like, and I had several problems. Not asking for a fix but a conversation or brainstorm.

1. Looked into the laps API and I found an issue with ocon's data for the monaco GP https://github.com/br-g/openf1/issues/69
2. The laps API seems more promising of a solution to me than the intervals API, however one issue I see is not knowing the total number of laps to compare to. My thought is to take the number of laps completed by the winner (which I already know from using the positions API) and use that as the basis for comparison.
3. I did some investigating with the intervals API but the previous proposed solution of using the laps API seems better/easier.

**Comment by michae1michae1** [comment]
I've been playing around with AI applications and I have been able to reliably identify DNF from radio messages...  It's actually pretty wild...

Basically I feed a small LLM the radio message data and.. ya it can do a lot of things pretty reliably... surprisingly

**Reply by mohamedhadrami** [reply]
Just curious, have you been playing around with this some more?

**Reply by michae1michae1** [reply]
Haven’t made any improvements lately - but most of the “work” is done tuning the prompts and use a model with function calling to make use of the responses

**Reply by mohamedhadrami** [reply]
Sweet, it would be a great to play around with once you get around to it

**Comment by rightpalava** [comment]
If you look at the laps method and look for the latest lap it will only show drivers who completed that lap, so you know that for example with the 2024 Chinese GP by Lap 50 we had 3 drivers DNF, looking at https://api.openf1.org/v1/laps?session_key=9673&lap_number=50 and you only get 17 entries returned.  That's how I'm managing anyway - obvs you'll need to know which lap to look at, there isn't a 'latest' option for 'lap_number'

**Reply by michae1michae1** [reply]
Ya accurately getting DNF is all about the edge cases.

also there’s a big difference in understanding DNF from historical data vs real time applications 

Strongly encourage utilizing radio messages (:
